theano/tensor/tests/test_sort.py

```diff
@@ -219,3 +235,247 @@ def test_argsort_grad():
+        idx = slice(-k, None) if k > 0 else slice(-k)
+        goali = np.argsort(xval)[idx].astype(idx_dtype)
+        goalv = xval[goali]
+
+        # due to uniqueness, we expect indices same
+        assert np.all(xval[np.sort(yival)] == xval[np.sort(goali)])
+        assert np.allclose(np.sort(yvval), goalv)
+
+    @utt.parameterized.expand(chain(
+        product(
+            (18, 62, 258),
+            (1, -1, 'n//2'),
+            ('int32', 'float32')),
+        ((2048, 1337, 'float32'),)))
+    def test_argtopk_1d_collision(self, size, k, dtype):
+        # with non-unique kth max value
+        if isinstance(k, str):
+            k = eval(k.replace('n', str(size)))
+
+        x = theano.tensor.vector(name='x', dtype=dtype)
+        y = argtopk(x, k, idx_dtype='int32')
+        fn = theano.function([x], y)
+        xval = np.repeat(np.random.uniform(-100., 100., size=size // 2).astype(dtype), 2)
+        xval = xval[np.random.permutation(size)]
+        yval = fn(xval)
+        idx = slice(-k, None) if k > 0 else slice(-k)
+        goal = np.argsort(xval)[idx].astype('int32')
+        print(goal)
+        print(np.argsort(xval))
+        assert np.allclose(np.sort(xval[yval]), np.sort(xval[goal]))
+
+    @utt.parameterized.expand(product(
+        ((1, 1), (2, 3), (17, 15), (15, 17), (11, 7, 5), (2, 3, 5, 7, 11), (2017, 5, 3)),
+        (1, -1, '(1+n)//2', 'n-1', '-n', '1-n'),
+        ('float32', 'int32'),
+        ('int32', 'int64')))
+    def test_argtopk_nd(self, shp, k_, dtype, idx_dtype):
+        ndim = len(shp)
+        for axis in range(-ndim, ndim):
+            if isinstance(k_, str):
+                k = eval(k_.replace('n', str(shp[axis])))
+            else:
+                k = k_
+
+            if k == 0:
+                continue
+
+            x = theano.tensor.tensor(
+                name='x', broadcastable=(False,) * len(shp), dtype=dtype)
+            y = argtopk(x, k, axis=axis, idx_dtype=idx_dtype)
+            fn = theano.function([x], y)
+            size = reduce(int.__mul__, shp)
+            xval = gen_unique_vector(size, dtype).reshape(shp)
+            yval = fn(xval)
+            idx = slice(-k, None) if k > 0 else slice(-k)
+            l = axis % ndim
+            r = ndim - l
+            idx = (slice(None),) * l + (idx,) + (slice(None),) * (r - 1)
+            goal = np.argsort(xval, axis=axis)[idx].astype(idx_dtype)
+
+            print(dict(k=k, axis=axis, shp=shp))
+            print('x:')
+            print(xval)
+            print('y:')
+            print(np.sort(yval, axis=axis))
+            print('goal:')
+            print(np.sort(goal, axis=axis))
+            # print(np.argsort(xval))
+            assert np.all(np.sort(yval, axis=axis) == np.sort(goal, axis=axis))
+
+
+class TopKInferShapeTester(utt.InferShapeTester):
+    @utt.parameterized.expand(product(
+        ((2, 3), (15, 17), (11, 7, 5), (2, 3, 5, 7, 11), (2, 4, 3, 1)),
+        (1, -1, '(1+n)//2', 'n-1', '-n', '1-n')))
```

Comment:
You can test less cases here. In the first row, keep just 3 shapes with different ndim. On the second row, keep just 3 values.

idem bellow